crates/ra_ide/src/hover.rs

```diff
@@ -215,6 +288,77 @@ fn hover_text_from_name_kind(db: &RootDatabase, def: Definition) -> Option<Strin
     }
 }
 
+fn goto_type_action(db: &RootDatabase, def: Definition) -> Option<HoverAction> {
+    fn add_subst_targets(
+        db: &RootDatabase,
+        substs: Option<&Substs>,
+        targets: &mut Vec<HoverGotoTypeData>,
+    ) {
+        if let Some(substs) = substs {
+            for ty in substs.iter() {
+                if let Some((adt_id, adt_params)) = ty.strip_references().as_adt() {
+                    let adt = Adt::from(adt_id);
+                    let mod_path = adt_mod_path(db, &adt);
+                    let item = HoverGotoTypeData::new(mod_path, adt.to_nav(db));
+                    if !targets.contains(&item) {
+                        targets.push(item);
+                    }
+                    add_subst_targets(db, Some(adt_params), targets);
+                }
+            }
+        }
+    }
+
+    fn to_impl_action(nav_target: NavigationTarget) -> HoverAction {
+        HoverAction::Implementaion(FilePosition {
+            file_id: nav_target.file_id(),
+            offset: nav_target.range().start(),
+        })
+    }
+
+    match def {
+        Definition::Local(it) => {
+            let ty = it.ty(db);
+            let adt: Adt = ty.autoderef(db).find_map(|ty| ty.as_adt())?;
+            let mod_path = adt_mod_path(db, &adt);
+            let mut targets = vec![HoverGotoTypeData::new(mod_path, adt.to_nav(db))]; // not a Set to preserve the order
+            add_subst_targets(db, ty.substs().as_ref(), &mut targets);
+
+            Some(HoverAction::GoToType(targets))
+        }
+        Definition::ModuleDef(it) => match it {
+            ModuleDef::Adt(Adt::Struct(it)) => Some(to_impl_action(it.to_nav(db))),
+            ModuleDef::Adt(Adt::Union(it)) => Some(to_impl_action(it.to_nav(db))),
+            ModuleDef::Adt(Adt::Enum(it)) => Some(to_impl_action(it.to_nav(db))),
+            ModuleDef::Trait(it) => Some(to_impl_action(it.to_nav(db))),
+            _ => None,
+        },
+        _ => None,
+    }
+}
```

Comment:
I feel this is the most complex action here. Would you mind moving it to a separate PR, so that we first can review the general LSP interface changes, and then tackle the specifics of type infernce and such? As per our would-be-introduce contribution guidelines :-) 

https://github.com/rust-analyzer/rust-analyzer/pull/4703/files#diff-83b898006efcb044dce73a8a5cc94ed6R156-R161

Reply:
Sure, I'll close this PR and split it to a several ones.